catalog/plugin/catalog-plugin-configuration-mapping/src/main/java/org/codice/ddf/catalog/plugin/configuration/mapping/ConfigurationMappingPlugin.java

```diff
@@ -0,0 +1,143 @@
+
+import ddf.catalog.data.Attribute;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.data.impl.AttributeImpl;
+import ddf.catalog.operation.CreateRequest;
+import ddf.catalog.operation.DeleteRequest;
+import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.plugin.PluginExecutionException;
+import ddf.catalog.plugin.PreIngestPlugin;
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.security.expansion.Expansion;
+
+public class ConfigurationMappingPlugin implements PreIngestPlugin {
+
+    private Expansion expansionService;
+
+    @Override
+    public CreateRequest process(CreateRequest input)
+            throws PluginExecutionException, StopProcessingException {
+
+        expandAttributes(input.getMetacards());
+        return input;
+    }
+
+    @Override
+    public UpdateRequest process(UpdateRequest input)
+            throws PluginExecutionException, StopProcessingException {
+
+        List<Metacard> metacards = new ArrayList<>();
+
+        for (Map.Entry<Serializable, Metacard> metacardEntry : input.getUpdates()) {
+            metacards.add(metacardEntry.getValue());
+        }
+
+        expandAttributes(metacards);
+        return input;
+    }
+
+    @Override
+    public DeleteRequest process(DeleteRequest input)
+            throws PluginExecutionException, StopProcessingException {
+        return input;
+    }
+
+    private void expandAttributes(List<Metacard> metacards) {
+
+        AttributeRuleSet attributeRuleSet = new AttributeRuleSet(expansionService);
+
+        for (Metacard metacard : metacards) {
+            attributeRuleSet.expandMetacard(metacard);
+        }
+    }
+
+    private class AttributeRuleSet {
+
+        private Map<String, Map<String, String>> attributeRuleSet;
+
+        AttributeRuleSet(Expansion expansion) {
+
+            attributeRuleSet = new HashMap<>();
+
+            for (Map.Entry<String, List<String[]>> attributeRules : expansion.getExpansionMap()
+                    .entrySet()) {
+                Map<String, String> attributeMap = toMap(attributeRules.getValue());
+                attributeRuleSet.put(attributeRules.getKey(), attributeMap);
+            }
+        }
+
+        void expandMetacard(Metacard metacard) {
+            Set<String> attributeSet = metacard.getMetacardType()
+                    .getAttributeDescriptors()
+                    .stream()
+                    .map(des -> des.getName())
+                    .collect(Collectors.toSet());
+
+            Set<String> commonAttributes = new HashSet<>(CollectionUtils.intersection(attributeSet,
+                    attributeRuleSet.keySet()));
+
+            for (String attributeName : commonAttributes) {
+                Attribute metacardAttribute = metacard.getAttribute(attributeName);
+
+                if (metacardAttribute != null) {
+                    HashSet attributeValues = new HashSet<String>();
+                    attributeValues.addAll(metacardAttribute.getValues());
```

Comment:
I feel like we should still add a test to verify that given a multi-valued attribute, only the value matching the expansion gets replaced. It looks good to me though 👍